app/src/main/java/com/rkm/tasky/network/repository/implementation/AuthenticationRepositoryImpl.kt

```diff
@@ -0,0 +1,105 @@
+package com.rkm.tasky.network.repository.implementation
+
+import com.rkm.tasky.di.IoDispatcher
+import com.rkm.tasky.network.datasource.TaskyRemoteDataSource
+import com.rkm.tasky.network.model.dto.AccessTokenDTO
+import com.rkm.tasky.network.model.dto.LoginDTO
+import com.rkm.tasky.network.model.request.AccessTokenRequest
+import com.rkm.tasky.network.model.request.LoginUserRequest
+import com.rkm.tasky.network.model.request.RegisterUserRequest
+import com.rkm.tasky.network.model.response.asAccessTokenDTO
+import com.rkm.tasky.network.model.response.asLoginDTO
+import com.rkm.tasky.network.repository.abstraction.AuthenticationRepository
+import com.rkm.tasky.network.util.NetworkError
+import com.rkm.tasky.util.result.EmptyResult
+import com.rkm.tasky.util.result.Result
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import java.io.IOException
+import javax.inject.Inject
+
+class AuthenticationRepositoryImpl @Inject constructor(
+    private val dataSource: TaskyRemoteDataSource,
+    @IoDispatcher private val dispatcher: CoroutineDispatcher
+): AuthenticationRepository {
+    override suspend fun registerUser(
+        fullName: String,
+        email: String,
+        password: String
+    ): EmptyResult<NetworkError.AuthError> = withContext(dispatcher) {
+        return@withContext try {
+            val response = dataSource.registerUser(RegisterUserRequest(fullName, email, password))
+            if (response.isSuccessful) {
+               Result.Success(Unit)
+            } else {
+                Result.Error(NetworkError.AuthError.REGISTER_USER_ERROR)
+            }
+        } catch(exception: IOException) {
+            exception.asAuthError()
+        }
```

Comment:
Since you'll have a lot of these API calls that can all give you similar errors, could it make sense extracting try/catch into a function? Something like this maybe?

```
suspend inline fun <reified T> safeCall(execute: () -> HttpResponse): Result<T, DataError.Network> {
    val response = try {
        execute()
    } catch (e: UnresolvedAddressException) {
        e.printStackTrace()
        return Result.Error(DataError.Network.NO_INTERNET)
    } catch (e: HttpRequestTimeoutException) {
        e.printStackTrace()
        return Result.Error(DataError.Network.REQUEST_TIMEOUT)
    } catch (e: SerializationException) {
        e.printStackTrace()
        return Result.Error(DataError.Network.SERIALIZATION)
    } catch (e: Exception) {
        if (e is CancellationException) throw e
        e.printStackTrace()
        return Result.Error(DataError.Network.UNKNOWN)
    }

    return responseToResult(response)
}

suspend inline fun <reified T> responseToResult(response: HttpResponse): Result<T, DataError.Network> {
    return when (response.status.value) {
        in 200..299 -> Result.Success(response.body<T>())
        401 -> Result.Error(DataError.Network.UNAUTHORIZED)
        408 -> Result.Error(DataError.Network.REQUEST_TIMEOUT)
        409 -> Result.Error(DataError.Network.CONFLICT)
        413 -> Result.Error(DataError.Network.PAYLOAD_TOO_LARGE)
        429 -> Result.Error(DataError.Network.TOO_MANY_REQUESTS)
        in 500..599 -> Result.Error(DataError.Network.SERVER_ERROR)
        else -> Result.Error(DataError.Network.UNKNOWN)
    }
}
```

Reply:
This makes much more sense. I think I know how to do what I originally wanted to achieve.